Enhancements to Improve Modeling of Data Centers

xuanluo113 opened the pull request:
Pull request overview
---------------------
Data centers in the United States consumed about 2% of the nation’s electricity power, and its energy use is growing faster than other building sectors. The proposed enhancements build upon existing capabilities of EnergyPlus to model energy performance of data centers (the ITE object) by enabling more flexibility to consider impact of air flow configurations and management strategies in data centers. The proposed enhancements were derived from discussions with members of the ASHRAE Standard 90.1/90.4, Technical Committee 9.9 and LBNL’s data center researchers.

New options have been added to ElectricEquipment:ITE:AirCooled to control supply air flow based on user-specified supply and return approach temperatures.

### Work Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
   - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
   - Refactoring: This pull request includes code changes that don't change the functionality of the program, just perform refactoring
   - NewFeature: This pull request includes code to add a new feature to EnergyPlus
   - Performance: This pull request includes code changes that are directed at improving the runtime performance of EnergyPlus
   - DoNoPublish: This pull request includes changes that shouldn't be included in the changelog

### Review Checklist
This will not be exhaustively relevant to every PR.
 - [x] Code style (parentheses padding, variable names)
 - [x] Functional code review (it has to work!)
 - [x] If defect, results of running current develop vs this branch should exhibit the fix
 - [ ] CI status: all green or justified
 - [ ] Performance: CI Linux results include performance check -- verify this
 - [x] Unit Test(s)
 - C++ checks:
   - [ ] Argument types
   - [ ] If any virtual classes, ensure virtual destructor included, other things
 - IDD changes:
   - [x] Verify naming conventions and styles, memos and notes and defaults
   - [x] Open windows IDF Editor with modified IDD to check for errors
   - [x] If transition, add rules to spreadsheet
   - [ ] If transition, add transition source
   - [x] If transition, update idfs
 - [x] If new idf included, locally check the err file and other outputs
 - [x] Documentation changes in place
 - [x] Changed docs build successfully
 n/a ExpandObjects changes?
 n/a If output changes, including tabular output structure, add to output rules file for interfaces

xuanluo113 committed:
NFP

xuanluo113 committed:
Data Center NFP

nrel-bot-2 commented:
@hongtz68 @XuanLuoLBL @lgentile it has been 16 days since this pull request was last updated.

nrel-bot-3 commented:
@hongtz68 @XuanLuoLBL @lgentile it has been 21 days since this pull request was last updated.

nrel-bot-3 commented:
@hongtz68 @XuanLuoLBL @lgentile it has been 14 days since this pull request was last updated.

nrel-bot-3 commented:
@hongtz68 @XuanLuoLBL @lgentile it has been 15 days since this pull request was last updated.

nrel-bot-2 commented:
@hongtz68 @XuanLuoLBL @lgentile it has been 14 days since this pull request was last updated.

xuanluo113 committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into ImproveModelingDataCenter

xuanluo113 committed:
Code and test

xuanluo113 committed:
Debug

xuanluo113 committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into ImproveModelingDataCenter

xuanluo113 committed:
Add error check

xuanluo113 commented:
@mjwitte Hi Michael, the code, with a unit test, is ready to be reviewed. Thanks a lot.

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -21255,17 +21267,18 @@ ElectricEquipment:ITE:AirCooled,
        \note Name of the supply air inlet node serving this ITE. Required if the
-       \note Air Node Connection Type = AdjustedSupply. Also required if reporting of
+       \note Air Node Connection Type = AdjustedSupply. Also required if 
+	   \note Calculation Method = FlowControlWithApproachTemperatures. Also required if reporting of
```
Tab.

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -21306,22 +21319,44 @@ ElectricEquipment:ITE:AirCooled,
-
+  N10, \field Supply Approach Temperature 
+       \note Either N10 or A20 is required if Calculation Method is set to FlowControlWithApproachTemperatures 
+	   \note The difference of the IT inlet temperature from the AHU supply air temperature
```
More tabs here and following.  Use spaces.

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -21146,15 +21146,27 @@ OtherEquipment,
-  A3 , \field Design Power Input Calculation Method
+  A3 , \field Calculation Method 
+       \note The entered calculation method is used to specify how the IT inlet temperature and zone 
+	   \note return air temperature. If FlowFromSystem is chosen, the zone is assumed to be well-mixed. 
```
Indentation problem - make sure IDD is all spaces, no tabs.

xuanluo113 replied:
Fixed here and other places.

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -21306,22 +21319,44 @@ ElectricEquipment:ITE:AirCooled,
        \default ITE-UPS
-
+  N10, \field Supply Approach Temperature 
+       \note Either N10 or A20 is required if Calculation Method is set to FlowControlWithApproachTemperatures 
```
Use field names in notes, not "N10", "A20" etc.  These fields should all have a note saying "This field is ignored when Calculation Method is FlowFromSystem."
For clarity, I would prefer that the names of these four fields be " . . . Temperature Difference . . . ".  Either way, the units should be deltaC.

xuanluo113 replied:
Fixed this.

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -21146,15 +21146,27 @@ OtherEquipment,
        \type object-list
        \object-list ZoneNames
-  A3 , \field Design Power Input Calculation Method
+  A3 , \field Calculation Method 
```
As I look at this now, "Calculation Method" is ambiguous, especially when the next field is "Design Power Input Calculation Method".  Would "Air Flow Caclulation Method" or "Temperature Calculation Method" be a better field name here?

xuanluo113 replied:
Modified.

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -21146,15 +21146,27 @@ OtherEquipment,
+	   \note If FlowControlWithApproachTemperatures is chosen, Supply and Return approach temperature 
+	   \note should be defined to indicate the temperature difference due to the air distribution. When
+	   \note FlowFromSystem is chosen, the inputs of Air Inlet Connection Type, Design Recirculation Fraction
+	   \note and Recirculation Function of Loading and Supply Temperature Curve Name are ignored. For multiple
```
Is this correct?  Aren't these fields are used with FlowFromSystem, ignored for FlowControlWithApproachTemperatures?  Also, each of these fields below should include a note stating something like "This field is only used when Calculation Method is FlowFromSystem".

xuanluo113 replied:
Right... Fixed this and added the notes.

mjwitte commented on src/EnergyPlus/InternalHeatGains.cc:
```diff
@@ -94,6 +94,7 @@
 #include <DataHVACGlobals.hh>
 #include <ElectricPowerServiceManager.hh>
 #include <ExteriorEnergyUse.hh>
+#include <map>
```
Minor detail - this would be better up at line 50.

xuanluo113 replied:
Checked.

mjwitte commented on src/EnergyPlus/InternalHeatGains.cc:
```diff
@@ -2336,8 +2337,26 @@ namespace InternalHeatGains {
+						ErrorsFound = true;
+					}
+				}
+				if ( Zone( ZoneITEq( Loop ).ZonePtr ).HasApproachTempToReturnAir && ( !ZoneITEq( Loop ).FlowControlWithApproachTemps ) ) {
```
This error will not trip if, for example, zone 1 has two ZoneITEq objects and the first one is FlowFromSystem and the second one if FlowControlWithApproachTemps.  I think you need to loop through the objects again after they are all processed.

xuanluo113 replied:
Modified.

mjwitte commented on src/EnergyPlus/InternalHeatGains.cc:
```diff
@@ -2336,8 +2337,26 @@ namespace InternalHeatGains {
 				ZoneITEq( Loop ).Name = AlphaName( 1 );
 				ZoneITEq( Loop ).ZonePtr = FindItemInList( AlphaName( 2 ), Zone );
 
-				// IT equipment design level calculation method.
```
Minor detail - this comment line got lost.

xuanluo113 replied:
Fixed.

mjwitte commented on src/EnergyPlus/InternalHeatGains.cc:
```diff
@@ -2364,60 +2383,60 @@ namespace InternalHeatGains {
 
 				} else {
-					ShowSevereError( RoutineName + CurrentModuleObject + "=\"" + AlphaName( 1 ) + "\", invalid " + cAlphaFieldNames( 3 ) + ", value  =" + AlphaName( 3 ) );
+					ShowSevereError( RoutineName + CurrentModuleObject + "=\"" + AlphaName( 1 ) + "\", invalid " + cAlphaFieldNames( 3 ) + ", value  =" + AlphaName( 4 ) );
```
cAlphaFieldNames( 3 ) should be cAlphaFieldNames( 4 ).

xuanluo113 replied:
Fixed this and the other places.

mjwitte commented on src/EnergyPlus/InternalHeatGains.cc:
```diff
@@ -2364,60 +2383,60 @@ namespace InternalHeatGains {
 				SchMax = 0.0;
 				if ( ZoneITEq( Loop ).OperSchedPtr == 0 ) {
-					ShowSevereError( RoutineName + CurrentModuleObject + "=\"" + AlphaName( 1 ) + "\", invalid " + cAlphaFieldNames( 4 ) + " entered=" + AlphaName( 4 ) );
+					ShowSevereError( RoutineName + CurrentModuleObject + "=\"" + AlphaName( 1 ) + "\", invalid " + cAlphaFieldNames( 4 ) + " entered=" + AlphaName( 5 ) );
```
cAlphaFieldNames( 5 )

mjwitte commented on src/EnergyPlus/InternalHeatGains.cc:
```diff
@@ -2435,111 +2454,138 @@ namespace InternalHeatGains {
 				} else {
 					ShowSevereError( RoutineName + CurrentModuleObject + ": " + AlphaName( 1 ) );
-					ShowContinueError( "Invalid " + cAlphaFieldNames( 10 ) + '=' + AlphaName( 10 ) );
+					ShowContinueError( "Invalid " + cAlphaFieldNames( 10 ) + '=' + AlphaName( 11 ) );
```
cAlphaFieldNames( 11 )

mjwitte commented on src/EnergyPlus/InternalHeatGains.cc:
```diff
@@ -2435,111 +2454,138 @@ namespace InternalHeatGains {
+					if ( !lAlphaFieldBlanks( 20 ) ) {
+						ZoneITEq( Loop ).SupplyApproachTempSch = GetScheduleIndex( AlphaName( 20 ) );
+					} else {
+						if ( ZoneITEq( Loop ).SupplyApproachTemp == 0 ) {
```
If zero is not a valid input, then the IDD default should not be zero and there should be  \minimum> 0.0   Is there a reasonable value to use as a default?  If not, then do not declare one and change this test to be if blank.

xuanluo113 replied:
I changed the test to be if blank.

mjwitte commented on src/EnergyPlus/InternalHeatGains.cc:
```diff
@@ -2435,111 +2454,138 @@ namespace InternalHeatGains {
+					if ( !lAlphaFieldBlanks( 21 ) ) {
+						ZoneITEq( Loop ).ReturnApproachTempSch = GetScheduleIndex( AlphaName( 21 ) );
+					} else {
+						if ( ZoneITEq( Loop ).ReturnApproachTemp == 0 ) {
```
Same comments here about zero.

mjwitte commented on src/EnergyPlus/InternalHeatGains.cc:
```diff
@@ -2592,6 +2638,7 @@ namespace InternalHeatGains {
 					SetupOutputVariable( "Zone ITE Fan Electric Power at Design Inlet Conditions", OutputProcessor::Unit::W, ZnRpt( ZoneITEq( Loop ).ZonePtr ).ITEqFanPowerAtDesign, "Zone", "Average", Zone( ZoneITEq( Loop ).ZonePtr ).Name );
 					SetupOutputVariable( "Zone ITE UPS Heat Gain to Zone Rate", OutputProcessor::Unit::W, ZnRpt( ZoneITEq( Loop ).ZonePtr ).ITEqUPSGainRateToZone, "Zone", "Average", Zone( ZoneITEq( Loop ).ZonePtr ).Name );
 					SetupOutputVariable( "Zone ITE Total Heat Gain to Zone Rate", OutputProcessor::Unit::W, ZnRpt( ZoneITEq( Loop ).ZonePtr ).ITEqConGainRateToZone, "Zone", "Average", Zone( ZoneITEq( Loop ).ZonePtr ).Name );
+					SetupOutputVariable( "Zone ITE Adjusted Return Air Temperature", OutputProcessor::Unit::W, ZnRpt( Loop ).ITEAdjReturnTemp, "Zone", "Average", Zone( Loop ).Name );
```
Two instances of `Loop` should be `ZoneITEq( Loop ).ZonePtr`.

xuanluo113 replied:
Modified.

mjwitte commented on src/EnergyPlus/InternalHeatGains.cc:
```diff
@@ -3912,30 +3962,47 @@ namespace InternalHeatGains {
 					WSupply = Node( SupplyNodeNum ).HumRat;
 				} else {
 					ShowFatalError( RoutineName + ": ElectricEquipment:ITE:AirCooled " + ZoneITEq( Loop ).Name );
 					ShowContinueError( "Air Inlet Connection Type = AdjustedSupply but no Supply Air Node is specified." );
```
This was my mistake long ago.  ShowContinueError will never be executed if it follows ShowFatalError.  These should be changed to ShowSevereError then ShowFatalError.  Same below.

xuanluo113 replied:
Changed.

mjwitte commented on src/EnergyPlus/ZoneEquipmentManager.cc:
```diff
@@ -4303,7 +4303,9 @@ namespace ZoneEquipmentManager {
-
+				if ( Zone( ActualZoneNum ).HasApproachTempToReturnAir && !( DataGlobals::BeginSimFlag ) ) {
+					Node( ReturnNode ).Temp = Zone( ActualZoneNum ).AdjustedTempToReturnAir;
+				}
```
This stomps on any other return air heat gain terms.  This should be moved up earlier (around line 4244) to reset TempZoneAir and TempRetAir.  Then other return air heat gain terms can be added to the base return temp from the IT Equipment.  Either that, or you need to check for all other possible return air heat gains (lights, airflow windows, refrigeration equipment, others?) and throw a fatal error is any of them are used in the same zone with IT equipment using the return approach method.

mjwitte commented on testfiles/1ZoneDataCenterCRAC_wPumpedDXCoolingCoil.idf:
```diff
@@ -846,6 +846,7 @@
   ElectricEquipment:ITE:AirCooled,
     Data Center Servers,     !- Name
     Main Zone,               !- Zone Name
+	,
```
No tabs.  Fill in with full field name and value, not blank here.  Same for 2ZoneDataCenterHVAC_wEconomizer.idf

xuanluo113 replied:
I added one testfile and modified another.

mjwitte commented on testfiles/2ZoneDataCenterHVAC_wEconomizer.idf:
```diff
@@ -2889,6 +2890,7 @@
   ElectricEquipment:ITE:AirCooled,
     EastDataCenter_Equip,    !- Name
     East Zone,               !- Zone Name
+	,
```
Actually, it would be good if one zone in this example file used the new model.  Or add a new example file that uses it.

mjwitte reviewed (commented):
@XuanLuoLBL Initial review comments follow.  I do not see any I/O Ref or Engineering Ref changes.  I have not built and tested the new feature yet.

xuanluo113 committed:
Addressing comments

xuanluo113 committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into ImproveModelingDataCenter

xuanluo113 committed:
Add a test file

xuanluo113 commented:
@mjwitte Mike, thanks for your review. I've addressed most of your comments, but while we were testing the feature with all kinds of cases, I got stuck in the return air flow problems.

Since we changed the return air temperature, as designed in the NFP, we're hoping to use the updated temperature to calculate the system with V_{AHU} = Q_{AHU}/(T_{return} - T_{supply}), but apparently the procedure is more complicated than the equation. For the previous commit, I realize changing the return node temp doesn't directly affect the V_AHU, but only reduces the DXCoil cooling load, which it doesn't suppose to. For this commit I was trying to handle this during sizing (as pointed in the file change with comments), but I'm not sure if this is right thing to d, and what is actually affected in the supply side.

An example of by-pass case is attached as a test file in the most recent commit.

Thanks for your help.

xuanluo113 commented on src/EnergyPlus/ZoneEquipmentManager.cc:
```diff
@@ -727,10 +727,16 @@ namespace ZoneEquipmentManager {
+						if ( DataHeatBalance::Zone( ActualZoneNum ).HasApproachTempToReturnAir && !( DataGlobals::BeginSimFlag ) ) {						
+							DeltaTemp = Temp - DataHeatBalance::Zone( ActualZoneNum ).AdjustedTempToReturnAir;
+						}
 						// If the user specify the design cooling supply air temperature difference, then
```
@mjwitte HERE.

mjwitte replied:
At first glance, these changes appear to be correct.  Do the sizing results for the zone cooling design flow rate change as expected?

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -21306,22 +21322,46 @@ ElectricEquipment:ITE:AirCooled,
+       \note Either Supply Temperature Difference or Supply Temperature Difference Schedule is required if Air Flow Calculation Method is set to FlowControlWithApproachTemperatures. 
+       \note This field is ignored when Air Flow Calculation Method is FlowFromSystem.
+       \type real
+       \units deltaC
```
minimum> 0 ??

xuanluo113 replied:
We'd assume these two temperatures can also be negative. As in the test files.

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -21306,22 +21322,46 @@ ElectricEquipment:ITE:AirCooled,
+       \note Either Return Temperature Difference or Return Temperature Difference Schedule is required if Air Flow Calculation Method is set to FlowControlWithApproachTemperatures.
+       \note This field is ignored when Air Flow Calculation Method is FlowFromSystem.
+       \type real
+       \units deltaC
```
minimum> 0 ??

mjwitte replied:
"unit outlet temperature" is not clear to me.  Is it correct to say "return air temperature" from the well mixed . . .?  Please clarify in the I/O Ref section also and in the next field.

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -21146,15 +21146,27 @@ OtherEquipment,
        \object-list ZoneNames
-  A3 , \field Design Power Input Calculation Method
+  A3 , \field Air Flow Caclulation Method
+       \note The entered calculation method is used to specify how the IT inlet temperature and zone 
```
This sentence is missing some words.  Suggest changing "The entered calculation method is used to specify how the IT . . ." to "The speicifed method is used to calculate the IT . . . "

xuanluo113 replied:
Modified. Thanks.

mjwitte commented on src/EnergyPlus/InternalHeatGains.cc:
```diff
@@ -3912,30 +3971,47 @@ namespace InternalHeatGains {
 				} else {
-					ShowFatalError( RoutineName + ": ElectricEquipment:ITE:AirCooled " + ZoneITEq( Loop ).Name );
+					ShowSevereError( RoutineName + ": ElectricEquipment:ITE:AirCooled " + ZoneITEq( Loop ).Name );
 					ShowContinueError( "Air Inlet Connection Type = AdjustedSupply but no Supply Air Node is specified." );
```
This line should be `ShowFatalError`, or set an errorflag here and fatal out later.

xuanluo113 replied:
Modified.

mjwitte commented on src/EnergyPlus/InternalHeatGains.cc:
```diff
@@ -3912,30 +3971,47 @@ namespace InternalHeatGains {
+						WSupply = Node( SupplyNodeNum ).HumRat;
+					} else {
+						ShowSevereError( RoutineName + ": ElectricEquipment:ITE:AirCooled " + ZoneITEq( Loop ).Name );
+						ShowContinueError( "Air Inlet Connection Type = AdjustedSupply but no Supply Air Node is specified." );
```
This line should be `ShowFatalError`, or set an errorflag here and fatal out later.

mjwitte commented on src/EnergyPlus/ZoneEquipmentManager.cc:
```diff
@@ -4292,6 +4298,11 @@ namespace ZoneEquipmentManager {
+						// Overwrite heat-to-return from ITE objects
+						if ( Zone( ActualZoneNum ).HasApproachTempToReturnAir && !( DataGlobals::BeginSimFlag ) ) {
+							TempRetAir = Zone( ActualZoneNum ).AdjustedTempToReturnAir;
+							Node( ReturnNode ).Temp = TempRetAir;
```
These lines still need to change, or add a check to disallow any return air heat gains with this type of ITE.

xuanluo113 replied:
We disabled other return air heat gains.

mjwitte reviewed (commented):
Some additional code comments below.  

Still need I/O Ref and Engineering Ref.  

Also, unit test InternalHeatGains_ElectricEquipITE_ApproachTemperatures is failing.

nrel-bot-3 commented:
@hongtz68 @XuanLuoLBL @lgentile it has been 15 days since this pull request was last updated.

xuanluo113 committed:
Fixed bugs and add documentation

xuanluo113 committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into ImproveModelingDataCenter

xuanluo113 commented:
@mjwitte 
We've addressed the last comments as the last commit. We've also added the documentation.

For the sizing issue, as we document in the engineering reference:

When FlowControlWithApproachTemperatures method is applied, during sizing and simulation, as the zone is assumed to be unevenly distributed, the supply air flow rate is calculated as:

{\dot m_{sys}} = {Q_{sys}}/({C_{p,air}} \cdot ({T_{in}} - {T_{return}})),

where T_{return} is the weighted return air temperature from all ITE objects in the zone. 

In this case, in the controlled zones with ITE objects, unlike the original thermostat setpoint control logic, the well-mixed zone temperature is not controlled. The actual controlled object is the supply air temperature. So when using this method, the user input of zone cooling setpoint is ignored and the unmet hour for cooling does not apply.

This is our current solution to make the sizing converge when overwriting the returning temperature. Wondering if that makes sense. For the current implementation:

1. the new model ignores the cooling setpoint of user input.
2. the unmet hour is calculated using the overridden setpoint.

Wondering what's your thought on this.

xuanluo113 committed:
Debug test case

xuanluo113 committed:
Try one more time

xuanluo113 committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into ImproveModelingDataCenter

nrel-bot-3 commented:
@hongtz68 @XuanLuoLBL @lgentile it has been 15 days since this pull request was last updated.

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into ImproveModelingDataCenter

mjwitte committed:
Merge branch 'ImproveModelingDataCenter' of https://github.com/NREL/EnergyPlus into ImproveModelingDataCenter

mjwitte commented on src/EnergyPlus/DataHeatBalance.hh:
```diff
@@ -1852,6 +1857,11 @@ namespace DataHeatBalance {
 		Real64 EMSUPSPower; // Value EMS is directing to use for override of UPS power [W]
+		Real64 SupplyApproachTemp; // The difference of the IT inlet temperature from the AHU supply air temperature
+		int SupplyApproachTempSch; // The difference schedule of the IT inlet temperature from the AHU supply air temperature
+		Real64 ReturnApproachTemp; // The difference of the unit outlet temperature from the well mixed zone temperature
```
Same comment here as yesterday - "unit outlet" is ambiguous - it's not the outlet temp from the IT equipment, it's the return temp for the return node flowing into the CRAC.
This raises another question - can this model coexist with another HVAC system in the same zone?  If so, the IT equipment needs an input for the return air node name.  If not, then it should throw an error if the zone has more than one return node?

xuanluo113 replied:
The change only applies for SingleDuct:VAV. We added check (warning) on if the HVAC system applies.

mjwitte commented on src/EnergyPlus/DataHeatBalance.hh:
```diff
@@ -1370,6 +1370,8 @@ namespace DataHeatBalance {
 		bool NoHeatToReturnAir; // TRUE means that heat to return air should be added to the zone load
 		bool RefrigCaseRA; // TRUE means there is potentially heat removal from return air
+		bool HasApproachTempToReturnAir; // TRUE means that return temp to return air is different from the zone mixed air temp
+		Real64 AdjustedTempToReturnAir; // Diff of the return temp from the zone mixed air temp
```
Based on how this is used, this variable is "Return temp adjusted by IT equipment model" or something like that.
Also, since these two variables are specific to zones with IT equipment, it seems these variables should include IT or ITE in the name and the description.

xuanluo113 replied:
We modified the term to AdjustedReturnTempByITE.

mjwitte commented on src/EnergyPlus/InternalHeatGains.cc:
```diff
@@ -2337,7 +2338,29 @@ namespace InternalHeatGains {
+							ShowSevereError( RoutineName + CurrentModuleObject + "=\"" + AlphaName( 1 ) + "\": invalid calculation method: " + AlphaName( 3 ) );
+							ShowContinueError( "Other return air heat gains from window or lights are not allowed when Air Flow Calculation Method = FlowControlWithApproachTemperatures." );
+							ErrorsFound = true;
+						}
```
This doesn't seem like the correct thing to check.  This code only executes once, so if the lights are off at that point, then this error won't fire.  May need to add a zone-level flag for HasLtsRetAirGain?  Better to find a place to check once somthing like
```
if ( Zone( ZoneNum ).HasApproachTempToReturnAir && (ZoneEquipConfig( ZoneEquipNum).ZoneHasAirFlowWindowReturn || Zone(  ZoneNum ).HasLtsRetAirGain ) then fatal error
```

xuanluo113 replied:
Indeed. We added the zone-level flag and a function to check zone return air after reading all heat gains.

mjwitte commented on src/EnergyPlus/ZoneEquipmentManager.cc:
```diff
@@ -4641,6 +4647,14 @@ namespace ZoneEquipmentManager {
+						if ( Zone( ActualZoneNum ).HasApproachTempToReturnAir && !( DataGlobals::BeginSimFlag ) ) {
+							if ( ZoneEquipConfig( ZoneNum ).ZoneHasAirFlowWindowReturn ) {
+								ShowFatalError( "Return air heat gains from window are not allowed when Air Flow Calculation Method = FlowControlWithApproachTemperatures." );
+							}
```
This test should only be done once.  See other comment where the lights return air heat gain error is checked.  There should be a single init location somewhere after all the appropriate input is processed where this can be checked once.

xuanluo113 replied:
As explained under the other comment, We checked the input after all inputs.

mjwitte commented on src/EnergyPlus/SingleDuct.cc:
```diff
@@ -2723,6 +2723,9 @@ namespace SingleDuct {
 			DeltaTemp = CpAirAvg * ( SysInlet( SysNum ).AirTemp - ZoneTemp );
+			if ( DataHeatBalance::Zone( ZoneNum ).HasApproachTempToReturnAir && !( DataGlobals::BeginSimFlag ) ) {
+				DeltaTemp = CpAirAvg * ( SysInlet( SysNum ).AirTemp - DataHeatBalance::Zone( ZoneNum ).AdjustedTempToReturnAir );
+			}
```
It appears that zones where `HasApproachTempToReturnAir` is true will only control properly with terminal units that call SimVAV, which is SingleDuct:VAV:Reheat and SingleDuct:VAV:NoReheat.  I don't see any error check for this or any mention of that in the I/O Ref or IDD notes.

Not sure exactly how this should be checked. You could warn for every equipment type that is not VAV, or scan the zone equipment to see if there is a VAV terminal in the zone and throw a severe if there isn't one.

mjwitte commented on src/EnergyPlus/ThermalComfort.cc:
```diff
@@ -2308,6 +2309,10 @@ namespace ThermalComfort {
+				
+				if ( Zone( iZone ).HasApproachTempToReturnAir ) {
+					deltaT = TempTstatAir( iZone ) - Zone( iZone ) .AdjustedTempToReturnAir;
+				}
```
This implies that the thermostat is now controlling on return air temperature.  That doesn't seem correct.

xuanluo113 replied:
As we discussed before. The room temperature will converge to the return air temperature for heat balance.

mjwitte commented on src/EnergyPlus/ZoneTempPredictorCorrector.cc:
```diff
@@ -3356,6 +3356,11 @@ namespace ZoneTempPredictorCorrector {
+			//
+			if ( DataHeatBalance::Zone( ZoneNum ).HasApproachTempToReturnAir && !( DataGlobals::BeginSimFlag ) ) {
+				LoadToCoolingSetPoint = TempDepZnLd( ZoneNum ) * DataHeatBalance::Zone( ZoneNum ).AdjustedTempToReturnAir - TempIndZnLd( ZoneNum );
+			}
```
This line (and similar ones below) are new?  If these lines are correct, then zones with HasApproachTempToReturnAir can't have *any* other HVAC equipment besides a single VAV terminal unit and the whole idea of what the load is changes.  Not sure about this.

mjwitte reviewed (commented):
More comments.  Some big questions about the thermostat setpoint and the load.

xuanluo113 committed:
Add Check Errors adn modify docs

xuanluo113 committed:
fix conflict

xuanluo113 committed:
fix conflict

xuanluo113 committed:
add check error

xuanluo113 commented:
@mjwitte Mike. Thanks for your detailed review. We modified the check of the errors and specified the system type. Also modified the docs and IDD. Please take a look if all make sense.

xuanluo113 commented:
@mjwitte The regression test failure regarding the 2 zone data center is expected, as we changed the idf.

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into ImproveModelingDataCenter

mjwitte committed:
Data center equipment - add transition rules

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into ImproveModelingDataCenter

mjwitte commented on src/EnergyPlus/InternalHeatGains.cc:
```diff
@@ -4096,6 +4219,31 @@ namespace InternalHeatGains {
 		}
 
+		std::map< int, std::vector< int > >::iterator it = ZoneITEMap.begin();
+		Real64 totalGain;
```
@Myoldmopar You ok with using a map here?

Myoldmopar replied:
It could be cleaned up to become a regular for loop instead of initializing begin() out here and then writing the while loop, but it doesn't make a significant difference.  At this point I'm OK with it.

mjwitte commented:
@Myoldmopar I'll merge this if CI comes back clean.  There's a followup issue #6480 to clean up a few things later.

Myoldmopar commented:
CI looks clean, just the one diff for data centers.  I'll go ahead and merge to get this one out of the way.
